src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -0,0 +1,884 @@
+                    int sourceIndex = 0;
+                    if (_state.BufferedCount != 0)
+                    {
+                        int loadSize = InternalBufferLengthBytes - (int)_state.BufferedCount;
+
+                        source.Slice(0, loadSize).CopyTo(new Span<byte>(buffer + _state.BufferedCount, loadSize));
+                        sourceIndex = loadSize;
+
+                        ConsumeStripes(accumulators, ref _state.NumStripesProcessed, NumStripesPerBlock, buffer, InternalBufferStripes, secret);
+                        _state.BufferedCount = 0;
+                    }
+                    Debug.Assert(sourceIndex < source.Length);
+
+                    // Large input to consume: ingest per full block.
+                    if (source.Length - sourceIndex > NumStripesPerBlock * StripeLengthBytes)
+                    {
+                        ulong stripes = (ulong)(source.Length - sourceIndex - 1) / StripeLengthBytes;
+                        Debug.Assert(NumStripesPerBlock >= _state.NumStripesProcessed);
+
+                        // join to current block's end
+                        ulong stripesToEnd = NumStripesPerBlock - _state.NumStripesProcessed;
+                        Debug.Assert(stripesToEnd <= stripes);
+                        Accumulate(accumulators, sourcePtr + sourceIndex, secret + ((int)_state.NumStripesProcessed * SecretConsumeRateBytes), (int)stripesToEnd);
+                        ScrambleAccumulators(accumulators, secret + (SecretLimitBytes));
+                        _state.NumStripesProcessed = 0;
+                        sourceIndex += (int)stripesToEnd * StripeLengthBytes;
+                        stripes -= stripesToEnd;
+
+                        // Consume per entire blocks.
+                        while (stripes >= NumStripesPerBlock)
+                        {
+                            Accumulate(accumulators, sourcePtr + sourceIndex, secret, NumStripesPerBlock);
+                            ScrambleAccumulators(accumulators, secret + (SecretLimitBytes));
+                            sourceIndex += NumStripesPerBlock * StripeLengthBytes;
+                            stripes -= NumStripesPerBlock;
+                        }
+
+                        // Consume last partial block.
+                        Accumulate(accumulators, sourcePtr + sourceIndex, secret, (int)stripes);
+                        sourceIndex += (int)stripes * StripeLengthBytes;
+                        Debug.Assert(sourceIndex < source.Length);  // at least some bytes left
+                        _state.NumStripesProcessed = stripes;
+
+                        // Buffer predecessor of last partial stripe.
+                        source.Slice(sourceIndex - StripeLengthBytes, StripeLengthBytes).CopyTo(new Span<byte>(buffer + InternalBufferLengthBytes - StripeLengthBytes, StripeLengthBytes));
+                    }
+                    else if (source.Length - sourceIndex > InternalBufferLengthBytes)
```

Comment:
I've just followed-suit to what the reference implementation is doing here.

If you think this can be simplified, I'd be happy to see that done, but let's leave that to a subsequent PR.